### ISSUE TITLE
parallel="classes" does not work if class extends abstract class ,but works fine parallel="tests" 

### ISSUE THREAD

**Issue by Jesse1321** [cbeust/testng#1334]
### TestNG Version:6.10
> Note: only the latest version is supported

### Expected behavior
should work

### Actual behavior
does not work. stopped at abstract class method.

### Is the issue reproductible on runner?

- [ ] Shell
- [X] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [X] NetBeans

### Test case sample
> Please, share the test case (as small as possible) which shows the issue

====testng.xml====

```xml
<?xml version='1.0' encoding='UTF-8' ?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="com.test_e2e_jar_0.0.1-SNAPSHOT" **parallel="classes"** thread-count="3" configfailurepolicy="continue">

    <!--
    see examples at http://testng.org/doc/documentation-main.html#testng-xml

    <suite-files>
        <suite-file path="./junit-suite.xml" />
    </suite-files>

    <test name="TimeOut">
        <classes>
            <class name="test.timeout.TimeOutTest" />
            <class name="test.timeout.TimeOutFromXmlTest"/>
            <class name="test.timeout.TimeOutThreadLocalSampleTest"/>
        </classes>
    </test>
    -->
    
    <test name="AutoNumberCategoryTest">
        <classes>
            <class name="com.test.e2e.testcase.common.AutoNumberCategoryTest" />
            <class name="com.test.e2e.testcase.common.AutoNumberingTest" />
            <class name="com.test.e2e.testcase.common.CodeTypeTest" />
            <class name="com.test.e2e.testcase.common.ContactTypeTest" />
            <class name="com.test.e2e.testcase.common.PropertyConditionTest" />
            <class name="com.test.e2e.testcase.common.PropertyTypeTest" />
            <class name="com.test.e2e.testcase.common.UOMTest" />
            <class name="com.test.e2e.testcase.common.UOMConversionTest" />
        </classes>
    </test>
</suite>
```
parallel mode works fine if specify parallel="tests" in testng.xml:

```xml
<?xml version='1.0' encoding='UTF-8' ?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="com_test_e2e_jar_0.0.1-SNAPSHOT" **## parallel="tests"** thread-count="3" configfailurepolicy="continue">

    <!--
    see examples at http://testng.org/doc/documentation-main.html#testng-xml

    <suite-files>
        <suite-file path="./junit-suite.xml" />
    </suite-files>

    <test name="TimeOut">
        <classes>
            <class name="test.timeout.TimeOutTest" />
            <class name="test.timeout.TimeOutFromXmlTest"/>
            <class name="test.timeout.TimeOutThreadLocalSampleTest"/>
        </classes>
    </test>
    -->
    
    <test name="AutoNumberCategoryTest">
        <classes>
            <class name="com.test.e2e.testcase.common.AutoNumberCategoryTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
      
    </test>
    <test name="AutoNumberingTest">
        <classes>
            <class name="com.test.e2e.testcase.common.AutoNumberingTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>
        </classes>

    </test>
    
    <!-- <test name="CodeReferencesTest">
        <classes>
            <class name="com.test.e2e.testcase.common.CodeReferencesTest" >
               <methods>
                    <include name="test0060_deleteAction" />
                </methods>
            </class>      
        </classes>
    </test>-->
    
    <test name="CodeTypeTest">
        <classes>
            <class name="com.test.e2e.testcase.common.CodeTypeTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>

    
     <test name="ContactTypeTest">
        <classes>
            <class name="com.test.e2e.testcase.common.ContactTypeTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>

    
<!--    <test name="LicenseTypeTest">
        <classes>
            <class name="com.test.e2e.testcase.common.LicenseTypeTest" >
            </class>      
        </classes>
    </test>-->

    
    <test name="PropertyConditionTest">
        <classes>
            <class name="com.test.e2e.testcase.common.PropertyConditionTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>

    
    <test name="PropertyTypeTest">
        <classes>
            <class name="com.test.e2e.testcase.common.PropertyTypeTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>

    
    <test name="UOMTest">
        <classes>
            <class name="com.test.e2e.testcase.common.UOMTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>

    
    <test name="UOMConversionTest">
        <classes>
            <class name="com.test.e2e.testcase.common.UOMConversionTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>
</suite>
```

**Comment by juherr** [comment]
@Jesse1321 Thanks for the report.
Does `parallel="classes"` work if you remove the `abstract class`?

**Comment by Jesse1321** [comment]
thanks Juherr for your response! 
i've spent 2 days on this issue, now i narrow down test scope to only parallel execute 2 tests and observed that if first method failed @Test(priority = 1), then execution order is abnormal , it does not follow up the annotation we defined, it will execute @Test(priority = 6) first, and then execute @Test(priority = 3), the other test methods are dependent on test1 method.

===========Test01.java===================================
package com.test.e2e.testcase.common.test1;
import org.testng.annotations.Test;


public class Test01 extends TestBase {    
    @Test(priority = 1)
    public void test0010_createAction() {
        System.out.println("Test01---test0010_createAction");
        throw new org.openqa.selenium.WebDriverException();

    }

    
    @Test(priority = 2,dependsOnMethods = { "test0010_createAction" })
    public void test0020_simpleSearch() {
        System.out.println("Test01---test0020_simpleSearch");
    }

        @Test(priority = 3)
    public void test0030_advancedSearch() {
        System.out.println("Test01---test0030_advancedSearch");

    }

   
    @Test(priority = 4,dependsOnMethods = { "test0010_createAction" })
    public void test0040_viewAction() {
        System.out.println("Test01---test0040_viewAction");
    }

    
    @Test(priority = 5,dependsOnMethods = { "test0010_createAction" })
    public void test0050_modifyAction() {
        System.out.println("Test01---test0050_modifyAction");

    }

    
    @Test(priority = 6)
    public void test0060_deleteAction() {
        System.out.println("Test01---test0060_deleteAction");

    
}
}
==============end===============================
========================Test02.java===========================
package com.test.e2e.testcase.common.test1;

import org.testng.annotations.Test;

public class Test02 extends TestBase {


    
    @Test(priority = 1)
    public void test0010_createAction() {
        System.out.println("Test02---test0010_createAction");

    }

    
    @Test(priority = 2,dependsOnMethods = { "test0010_createAction" })
    public void test0020_simpleSearch() {
        System.out.println("Test02---test0020_simpleSearch");
    }

        @Test(priority = 3)
    public void test0030_advancedSearch() {
        System.out.println("Test02---test0030_advancedSearch");

    }

   
    @Test(priority = 4,dependsOnMethods = { "test0010_createAction" })
    public void test0040_viewAction() {
        System.out.println("Test02---test0040_viewAction");
    }

    
    @Test(priority = 5,dependsOnMethods = { "test0010_createAction" })
    public void test0050_modifyAction() {
        System.out.println("Test02---test0050_modifyAction");

    }

    
    @Test(priority = 6)
    public void test0060_deleteAction() {
        System.out.println("Test02---test0060_deleteAction");

    
}
}
=============================end===============================

====================TestBase.java======================
package com.test.e2e.testcase.common.test1;
import java.net.MalformedURLException;
import java.net.URL;
import org.openqa.selenium.Platform; 
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.RemoteWebDriver;
import org.testng.annotations.AfterClass;
import org.testng.annotations.BeforeClass;
 
public abstract class TestBase {
 
 protected WebDriver driver;
    @BeforeClass
    public void setUp() throws MalformedURLException {
    String NodeURL="http://localhost:4444/wd/hub";
    DesiredCapabilities capa =DesiredCapabilities.chrome();
    capa.setBrowserName("chrome");
    capa.setPlatform(Platform.ANY);
     driver=new RemoteWebDriver(new URL(NodeURL),capa);
     driver.manage().window().maximize();

    }
 
    @AfterClass
    public void closeBrowser() {
        driver.quit();
 
    }
}

======================end=============================

===================TestNG.xml===========================
<?xml version='1.0' encoding='UTF-8' ?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="Parallel test runs" parallel="classes" thread-count="2">
 
    <test name="T_01">
        <classes>
            <class name="com.oracle.civic.test.e2e.testcase.common.test1.Test01" />
            <class name="com.oracle.civic.test.e2e.testcase.common.test1.Test02" />
        </classes>
    </test>
 

 
</suite>
=============================end=========================


=================output log======================
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running TestSuite
Feb 16, 2017 10:17:19 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Attempting bi-dialect session, assuming Postel's Law holds true on the remote end
Feb 16, 2017 10:17:19 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Attempting bi-dialect session, assuming Postel's Law holds true on the remote end
Feb 16, 2017 10:17:21 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: OSS
Feb 16, 2017 10:17:21 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: OSS
Test01---test0010_createAction
Test02---test0010_createAction
Test01---test0060_deleteAction
Test02---test0020_simpleSearch
Test02---test0030_advancedSearch
Test02---test0040_viewAction
Test02---test0050_modifyAction
Test02---test0060_deleteAction
Test01---test0030_advancedSearch
Tests run: 12, Failures: 1, Errors: 0, Skipped: 3, Time elapsed: 7.195 sec <<< FAILURE! - in TestSuite
test0010_createAction(com.test.e2e.testcase.common.test1.Test01)  Time elapsed: 0.054 sec  <<< FAILURE!
org.openqa.selenium.WebDriverException: 
Build info: version: '3.0.1', revision: '1969d75', time: '2016-10-18 09:49:13 -0700'
System info: host: 'testdb1', ip: '10.196.6.130', os.name: 'Windows Server 2012 R2', os.arch: 'amd64', os.version: '6.3', java.version: '1.8.0_101'
Driver info: driver.version: unknown
	at com.test.e2e.testcase.common.test1.Test01.test0010_createAction(Test01.java:16)


Results :

Failed tests: 
  Test01.test0010_createAction:16 WebDriver Build info: version: '3.0.1', revisi...

Tests run: 12, Failures: 1, Errors: 0, Skipped: 3

------------------------------------------------------------------------
BUILD FAILURE
------------------------------------------------------------------------

**Comment by Jesse1321** [comment]
if we change testng.xml to below, it works fine with expected execution order.

<suite name="Parallel test runs" parallel="classes" thread-count="3">
    <test name="T_01">
        <classes>
            <class name="com.oracle.civic.test.e2e.testcase.common.test1.Test01" />
        </classes>
    </test>
    <test name="T_02">
        <classes>
            <class name="com.oracle.civic.test.e2e.testcase.common.test1.Test02" />
        </classes>
    </test>
</suite>


===========output log============
INFO: Detected dialect: OSS
Test01---test0010_createAction
Test01---test0030_advancedSearch
Test01---test0060_deleteAction
Feb 16, 2017 11:42:06 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Attempting bi-dialect session, assuming Postel's Law holds true on the remote end
Feb 16, 2017 11:42:07 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: OSS
Test02---test0010_createAction
Test02---test0020_simpleSearch
Test02---test0030_advancedSearch
Test02---test0040_viewAction
Test02---test0050_modifyAction
Test02---test0060_deleteAction

========================

**Comment by krmahadevan** [comment]
@Jesse1321 - I am not able to understand what is the problem here. The classes are getting invoked in parallel.

Here's the samples 

**TestBase.java**
```java
import org.testng.ITestResult;
import org.testng.Reporter;
import org.testng.annotations.AfterClass;
import org.testng.annotations.BeforeClass;

public abstract class TestBase {
    @BeforeClass
    public void setUp() {
    }

    @AfterClass
    public void closeBrowser() {
    }

    String detail() {
        ITestResult result = Reporter.getCurrentTestResult();
        String cname = result.getTestClass().getRealClass().getName();
        String mname = result.getMethod().getMethodName();
        return " " + cname + "." + mname + "() running on Thread [" + Thread.currentThread().getId() + "]";
    }
}
```

**Test01.java**
```java
import org.testng.annotations.Test;

public class Test01 extends TestBase {
    @Test(priority = 1)
    public void test0010_createAction() {
        System.out.println("Test01---test0010_createAction" + detail());
        throw new RuntimeException("Simulating failure");
    }

    @Test(priority = 2, dependsOnMethods = {"test0010_createAction"})
    public void test0020_simpleSearch() {
        System.out.println("Test01---test0020_simpleSearch" + detail());
    }

    @Test(priority = 3)
    public void test0030_advancedSearch() {
        System.out.println("Test01---test0030_advancedSearch" + detail());
    }

    @Test(priority = 4, dependsOnMethods = {"test0010_createAction"})
    public void test0040_viewAction() {
        System.out.println("Test01---test0040_viewAction" + detail());
    }

    @Test(priority = 5, dependsOnMethods = {"test0010_createAction"})
    public void test0050_modifyAction() {
        System.out.println("Test01---test0050_modifyAction" + detail());
    }

    @Test(priority = 6)
    public void test0060_deleteAction() {
        System.out.println("Test01---test0060_deleteAction" + detail());
    }
}
```

**Test02.java**
```java
import org.testng.annotations.Test;

public class Test02 extends TestBase {

    @Test(priority = 1)
    public void test0010_createAction() {
        System.out.println("Test02---test0010_createAction" + detail());
    }

    @Test(priority = 2, dependsOnMethods = {"test0010_createAction"})
    public void test0020_simpleSearch() {
        System.out.println("Test02---test0020_simpleSearch" + detail());
    }

    @Test(priority = 3)
    public void test0030_advancedSearch() {
        System.out.println("Test02---test0030_advancedSearch" + detail());
    }

    @Test(priority = 4, dependsOnMethods = {"test0010_createAction"})
    public void test0040_viewAction() {
        System.out.println("Test02---test0040_viewAction" + detail());
    }

    @Test(priority = 5, dependsOnMethods = {"test0010_createAction"})
    public void test0050_modifyAction() {
        System.out.println("Test02---test0050_modifyAction" + detail());
    }

    @Test(priority = 6)
    public void test0060_deleteAction() {
        System.out.println("Test02---test0060_deleteAction" + detail());
    }
}
```
Now with the suite xml as below 
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="1334_Suite" verbose="2" parallel="classes" thread-count="10">
    <test name="1334_Tests">
        <classes>
            <class name="com.rationaleemotions.github.issue1334.Test01"/>
            <class name="com.rationaleemotions.github.issue1334.Test02"/>
        </classes>
    </test>
</suite>
```

Here's the output (trimmed version)
```
Test01---test0010_createAction com.rationaleemotions.github.issue1334.Test01.test0010_createAction() running on Thread [11]
Test01---test0060_deleteAction com.rationaleemotions.github.issue1334.Test01.test0060_deleteAction() running on Thread [14]
Test01---test0030_advancedSearch com.rationaleemotions.github.issue1334.Test01.test0030_advancedSearch() running on Thread [14]

Test02---test0010_createAction com.rationaleemotions.github.issue1334.Test02.test0010_createAction() running on Thread [12]
Test02---test0020_simpleSearch com.rationaleemotions.github.issue1334.Test02.test0020_simpleSearch() running on Thread [13]
Test02---test0030_advancedSearch com.rationaleemotions.github.issue1334.Test02.test0030_advancedSearch() running on Thread [13]
Test02---test0040_viewAction com.rationaleemotions.github.issue1334.Test02.test0040_viewAction() running on Thread [13]
Test02---test0050_modifyAction com.rationaleemotions.github.issue1334.Test02.test0050_modifyAction() running on Thread [13]
Test02---test0060_deleteAction com.rationaleemotions.github.issue1334.Test02.test0060_deleteAction() running on Thread [13]

SKIPPED: test0020_simpleSearch
SKIPPED: test0040_viewAction
SKIPPED: test0050_modifyAction
```

Now when using the below xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="1334_Suite" verbose="2" parallel="classes" thread-count="10">
    <test name="T_01">
        <classes>
            <class name="com.rationaleemotions.github.issue1334.Test01"/>
        </classes>
    </test>
    <test name="T_02">
        <classes>
            <class name="com.rationaleemotions.github.issue1334.Test02"/>
        </classes>
    </test>
</suite>
```
Here's the output (trimmed version)
```
Test01---test0010_createAction com.rationaleemotions.github.issue1334.Test01.test0010_createAction() running on Thread [11]
Test01---test0030_advancedSearch com.rationaleemotions.github.issue1334.Test01.test0030_advancedSearch() running on Thread [13]
Test01---test0060_deleteAction com.rationaleemotions.github.issue1334.Test01.test0060_deleteAction() running on Thread [16]

Test02---test0010_createAction com.rationaleemotions.github.issue1334.Test02.test0010_createAction() running on Thread [17]
Test02---test0020_simpleSearch com.rationaleemotions.github.issue1334.Test02.test0020_simpleSearch() running on Thread [18]
Test02---test0030_advancedSearch com.rationaleemotions.github.issue1334.Test02.test0030_advancedSearch() running on Thread [19]
Test02---test0040_viewAction com.rationaleemotions.github.issue1334.Test02.test0040_viewAction() running on Thread [20]
Test02---test0050_modifyAction com.rationaleemotions.github.issue1334.Test02.test0050_modifyAction() running on Thread [21]
Test02---test0060_deleteAction com.rationaleemotions.github.issue1334.Test02.test0060_deleteAction() running on Thread [22]


SKIPPED: test0020_simpleSearch
SKIPPED: test0040_viewAction
SKIPPED: test0050_modifyAction

```

I can for sure say that the abstract class has got nothing to do with parallel executions (as seen from the first suite xml output)

**Comment by Jesse1321** [comment]
Thanks krmahadevan for your analysis!
the problem is execution order is not as expected with the suite xml as below
`<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="1334_Suite" verbose="2" parallel="classes" thread-count="10">
    <test name="1334_Tests">
        <classes>
            <class name="com.rationaleemotions.github.issue1334.Test01"/>
            <class name="com.rationaleemotions.github.issue1334.Test02"/>
        </classes>
    </test>
</suite>`
Test01---test0030_advancedSearch (priority = 3)<----we want this be executed before than deleteAction
Test01---test0060_deleteAction (priority = 6) 

in this case it executes Test01---test0060_deleteAction (priority = 6) before than Test01---test0030_advancedSearch (priority = 3), that is not expected execution order, we want it executes Test01---test0030_advancedSearch (priority = 3) before than Test01---test0060_deleteAction (priority = 6) , otherwise advancedSearch method will be failed as the entry is deleted.

Test01---test0010_createAction com.rationaleemotions.github.issue1334.Test01.test0010_createAction() running on Thread [11]
Test01---test0060_deleteAction com.rationaleemotions.github.issue1334.Test01.test0060_deleteAction() running on Thread [14]
Test01---test0030_advancedSearch com.rationaleemotions.github.issue1334.Test01.test0030_advancedSearch() running on Thread [14]

the execution order is correct if we use below suite xml:
`<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="1334_Suite" verbose="2" parallel="classes" thread-count="10">
    <test name="T_01">
        <classes>
            <class name="com.rationaleemotions.github.issue1334.Test01"/>
        </classes>
    </test>
    <test name="T_02">
        <classes>
            <class name="com.rationaleemotions.github.issue1334.Test02"/>
        </classes>
    </test>
</suite>`

thanks again and appreciate your hard work on testing these codes!

**Comment by krmahadevan** [comment]
@juherr My PR https://github.com/cbeust/testng/pull/1361 fixed this issue. But I have still gone ahead and raised a new PR (https://github.com/cbeust/testng/pull/1354)  that merely adds tests to validate this issue

**Comment by krmahadevan** [comment]
@Jesse1321 - The issue should be fixed now. Please try using the latest snapshot for TestNG and confirm.

**Comment by Jesse1321** [comment]
it works fine now. thanks for helping!
